controlpanel/api/aws.py

```diff
@@ -24,248 +22,415 @@ def iam_arn(resource, account=settings.AWS_ACCOUNT_ID):
-        :param tags: Tags {key: value} e.g. {'buckettype': 'datawarehouse'}
-        :type name: str
-        :type tags: dict
-        """
-        self._do(
-            "s3",
-            "put_bucket_tagging",
-            Bucket=name,
-            Tagging={
-                "TagSet": [{"Key": str(k), "Value": str(v)} for k, v in tags.items()]
-            },
-        )
+    for policy in role.policies.all():
+        policy.delete()
+
+    role.delete()
 
-    def put_public_access_block(
-        self,
-        bucket_name,
-        block_public_acls=True,
-        ignore_public_acls=True,
-        block_public_policy=True,
-        restrict_public_buckets=True,
-    ):
-        self._do(
-            "s3",
-            "put_public_access_block",
+
+def create_bucket(bucket_name, is_data_warehouse=False):
+    try:
+        bucket = boto3.resource('s3').create_bucket(
             Bucket=bucket_name,
-            PublicAccessBlockConfiguration={
-                "BlockPublicAcls": block_public_acls,
-                "IgnorePublicAcls": ignore_public_acls,
-                "BlockPublicPolicy": block_public_policy,
-                "RestrictPublicBuckets": restrict_public_buckets,
+            ACL='private',
+            CreateBucketConfiguration={
+                'LocationConstraint': settings.BUCKET_REGION,
             },
         )
-
-    def _get_inline_policy_document(self, name, resource_type, policy_name):
-        if not self.enabled:
-            return None
-
-        aws_kwargs = {
-            f"{resource_type.capitalize()}Name": name,
-            "PolicyName": policy_name,
-        }
-
+        if is_data_warehouse:
+            bucket.Tagging().put(Tagging={'TagSet': [{
+                'Key': 'buckettype',
+                'Value': 'datawarehouse',
+            }]})
+
+    except bucket.meta.client.exceptions.BucketAlreadyOwnedByYou:
+        log.warning(f'Skipping creating Bucket {bucket_name}: Already exists')
+        return
+
+    bucket.Logging().put(BucketLoggingStatus={'LoggingEnabled': {
+        'TargetBucket': settings.LOGS_BUCKET_NAME,
+        'TargetPrefix': f"{bucket_name}/",
+    }})
+    bucket.meta.client.put_bucket_encryption(
+        Bucket=bucket_name,
+        ServerSideEncryptionConfiguration={'Rules': [{
+            'ApplyServerSideEncryptionByDefault': {
+                'SSEAlgorithm': 'AES256',
+            },
+        }]},
+    )
+    bucket.meta.client.put_public_access_block(
+        Bucket=bucket_name,
+        PublicAccessBlockConfiguration={
+            'BlockPublicAcls': True,
+            'IgnorePublicAcls': True,
+            'BlockPublicPolicy': True,
+            'RestrictPublicBuckets': True,
+        },
+    )
+    return bucket
+
+
+class S3AccessPolicy:
+    """Provides a convenience wrapper around a RolePolicy object"""
+
+    def __init__(self, policy):
+        self.policy = policy
+        self._policy_document = None
+        self.statements = {}
+
+        self._load()
```

Comment:
This is a bit confusing but I haven't followed the various iterations closely so maybe I'm missing something...

The `policy` argument is always passed and it's coming from the IAM role inline policy (usually `role.Policy('s3-access')` or something like that).

So what's with all these weird instance attributes that we keep around?

- We store the policy passed to the constructor in `self.policy`
- `self. _policy_document ` is `None` but it would eventually contain the same thing after calling the `policy_document` property right?
- and we also have the `policy_document` property which is memoizing the same thing in `self._policy_document`

The `self.policy` attribute is a `boto` thing and we also use it to update the policy. Maybe we can get rid of some stuff and have something like:

```python
class S3AccessPolicy:

    def __init__(self, policy):
        self.policy = policy
        self.policy_document = policy.policy_document # do we need to make a copy?
        self.statements = {}
```

Would you still need the `self._policy_document` instance attribute and the `policy_document` property in that case?

Reply:
This is a bit ugly, I agree. The `role.Policy.policy_document` is a lazy-loaded property, so if we assign it to `self.policy_document` in the constructor, that will trigger the API call. That would mean we have to move the `_load()` method code into the constructor to catch a possible `NoSuchEntityException`, and we would have to duplicate in the `ManagedS3AccessPolicy` class.
I'll try again.

Reply:
Also, because the `role.Policy.policy_document` is the return value of an API call, it is a different object every time, so I store it in `self._policy_document`.